clients/drcachesim/tests/raw2trace_io.cpp

```diff
@@ -0,0 +1,161 @@
+ * It uses ptrace to make sure raw2trace only interacts with the filesystem when expected
+ * -- it doesn't open/close any files outside of mapping modules (which should be local).
+ *
+ * XXX: We use ptrace as opposed to running under memtrace with replaced file operations
+ * because raw2trace uses drmodtrack, which doesn't isolate under static memtrace.
+ */
+
+#include "droption.h"
+#include "tracer/raw2trace.h"
+#include "tracer/raw2trace_directory.h"
+#include <cassert>
+#include <errno.h>
+#include <fcntl.h>
+#include <iostream>
+#include <string>
+#include <sys/ptrace.h>
+#include <sys/user.h>
+#include <sys/wait.h>
+#include <syscall.h>
+#include <unistd.h>
+
+static droption_t<std::string> op_indir(DROPTION_SCOPE_FRONTEND, "indir", "",
+                                        "[Required] Directory with trace input files",
+                                        "Specifies a directory with raw files.");
+
+static droption_t<std::string> op_out(DROPTION_SCOPE_FRONTEND, "out", "",
+                                      "[Required] Path to output file",
+                                      "Specifies the path to the output file.");
+#if defined(X64)
+# define SYS_NUM(r) (r).orig_rax
+#elif defined(X86)
+# define SYS_NUM(r) (r).orig_eax
+#else
+# error "Test only supports x86"
+#endif
+
+int
+main(int argc, const char *argv[])
+{
+    std::string parse_err;
+    if (!droption_parser_t::parse_argv(DROPTION_SCOPE_FRONTEND, argc, (const char **)argv,
+                                       &parse_err, NULL) ||
+        op_indir.get_value().empty() || op_out.get_value().empty()) {
+        std::cerr << "Usage error: " << parse_err << "\nUsage:\n"
+                  << droption_parser_t::usage_short(DROPTION_SCOPE_ALL);
+        return 1;
+    }
+
+    /* Open input/output files outside of traced region. And explicitly don't destroy dir,
+     * so they never get closed.
+     */
+    raw2trace_directory_t *dir =
+        new raw2trace_directory_t(op_indir.get_value(), op_out.get_value());
+
+    pid_t pid = fork();
+    if (pid == -1) {
+        std::cerr << "Fork failed\n";
+        return 1;
+    }
+    if (pid != 0) {
+        /* parent */
+        long res;
+        int status;
+        while (true) {
+            int wait_res = waitpid(pid, &status, __WALL);
+            if (wait_res < 0) {
+                if (errno == EINTR)
+                    continue;
+                perror("Failed waiting");
+                return 1;
+            }
+            if (WIFEXITED(status))
+                break;
+            user_regs_struct regs;
+            res = ptrace(PTRACE_GETREGS, pid, NULL, &regs);
+            if (res < 0) {
+                perror("ptrace failed");
+                return 1;
+            }
+            /* We don't distinguish syscall entry/exit because orig_rax is set
+             * on both. */
+            if (SYS_NUM(regs) == __NR_open || SYS_NUM(regs) == __NR_openat ||
+                SYS_NUM(regs) == __NR_creat) {
```

Comment:
In cmake `UNIX` includes MacOS where there is no __NR_creat.  If it's simpler you could just exclude in cmake (and ifdef sanity check at top of this file) via `NOT APPLE`.